[C++][Python] Allow FileSystem::DeleteDirContents to succeed if the directory is missing

westonpace opened the pull request:
Also changes the dataset writer to use this new method (specifying `missing_dir_ok=true`).  This should address behavior seen in ARROW-12358

pitrou reviewed (commented):
Perhaps you can add to the generic filesystems tests in `arrow/filesystem/test_util.cc`?

jorisvandenbossche commented on python/pyarrow/fs.py:
```diff
@@ -346,18 +346,24 @@ def create_dir(self, path, recursive):
                 self.fs.rm(subpath)
 
-    def delete_dir_contents(self, path):
+    def delete_dir_contents(self, path, missing_dir_ok):
```
I suppose this will break existing implementations, because our C++ code will now pass this keyword, while existing handlers won't yet accept that keyword in this method like above. 

Now, that will _always_ happen if we add new keywords, and we want to be able to add new keywords (so not saying we shouldn't do this change). But therefore wondering:

- Should we recommend adding `**kwargs` (that get ignored) to all methods on the Handler class to avoid such breakage? (of course, that would ignore the keyword initially when we add one, so also result in unexpected behaviour for the user. So that is maybe not better as initially raising an error until the handler gets updated for the latest arrow version)
- In theory we could only pass the keyword in the C++ code (in `PyFileSystem::DeleteDirContents`) to the handler method _if_ it has a non-default value, and otherwise not pass it (and set a default on the line above). Of course that complicates the C++ code with an if/else block (and introduces a risk of introducing inconsistent defaults at some point if one would change).

pitrou replied:
I don't think either approach you suggest is better than the explicit compatibility breakage. This is indeed an annoyance for third-party implementations, but there are probably not many of them, and their authors should be competent enough to write the necessary compatibility handling code if required (basically by inspecting the PyArrow version).

westonpace commented:
> Perhaps you can add to the generic filesystems tests in arrow/filesystem/test_util.cc?

@pitrou 

Good idea, thanks.  I removed the test I had added to `localfs_test.cc` since this was a brand new test and a subset of the generic test.  I left the additions to `s3fs_test.cc` and `hdfs_test.cc` as they already had independent tests for DeleteDirContents.

westonpace commented:
CI was failing because I did not have a GCS solution.  I've added that and think this is probably ready.

pitrou commented on cpp/src/arrow/filesystem/gcsfs.cc:
```diff
@@ -477,10 +478,22 @@ class GcsFileSystem::Impl {
       submitted.push_back(DeferNotOk(io_context.executor()->Submit(async_delete, o)));
     }
 
+    if (!missing_dir_ok && !at_least_one_obj) {
+      // If any files were found the directory implicitly exists.  If none were
+      // found then we still have to consider a marker-only directory so we test
+      // for that.
+      ARROW_ASSIGN_OR_RAISE(auto file_info, GetFileInfo(p));
+      if (file_info.type() == FileType::NotFound) {
+        return Status::IOError("No directory or any files matching path: ", p.full_path);
+      }
+    }
```
Hmm, I think you can reuse the `dir` variable introduced above?
```suggestion
    if (!missing_dir_ok && !at_least_one_obj && !dir) {
      // No files were found and no directory marker exists
      return Status::IOError("No such directory: ", p.full_path);
    }
```

pitrou reviewed (commented):
Thanks! Just a suggestion, otherwise LGTM
